torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1568,22 +1515,22 @@ void ProcessGroupNCCL::watchdogHandler() {
               globalStore_->set(std::string(TIMEOUT_DUMP), vec);
             }
 
-            auto wakeUpTime = getWakeupTime(waitTimeoutDumpInMilSec_);
             if (dumpOnTimeout_ && !optAsyncDebugDump) {
               // Store debug info to storage. (By default to local disk)
-              optAsyncDebugDump = launchAsyncDebugDump();
+              optAsyncDebugDump = std::async(std::launch::async, [this]() {
+                return this->dumpDebuggingInfo();
+              });
+              // wait for the dump until timeout
+              waitForFutureOrTimeout(
+                  *optAsyncDebugDump,
+                  std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
+                  "Flight recorder dump in watchdog");
```

Comment:
I think we're missing the extra 2 min wait in this path? 

Also I think it would be better if we only had one path. Perhaps in another PR, refactor to let watchdog simply signal timeout occurred, and monitor thread to then exercise the dumping logic?

Reply:
Agree, checking the code path, watchdog thread can throw an exception immediately after detecting the timeout. will add a sleep here for now. Also agree that we probably should not dump from 2 different places. But kind of understand it why it is there now, watchdog thread to detect timeout on local rank, monitoring thread to detect timeout on remote ranks...